### ISSUE TITLE
Allegro locks up if display is not focused on startup

### ISSUE THREAD

**Issue by fatcerberus** [liballeg/allegro5#852]
If an Allegro-based game happens to get started with its window out of focus (this sometimes happens when starting programs from the command line, for example) and the window is not given focus within a few seconds, the process will lock up.  If this happens the window's close button starts flickering like crazy and can't be clicked - the only way to close it at this point is by forcibly terminating the process.

This is a long-standing issue and I've been unable to track down the cause thus far.

**Comment by dos1** [comment]
On which platform?

**Comment by fatcerberus** [comment]
Windows.

**Comment by tbass** [comment]
I have a similar problem occuring on MacOS but I can not reliably reproduce it.
It happens usually while starting to debug and the focus of the window is (probably) not given fast enough from QtCreator (IDE I am using) to my game window. Then I will only see a black screen and the program locks up and crashes.
I do not have any flickering window buttons or similar.
Right now I am not at home, so I cannot tell the correct location where Allegro drops out. I will update when I have more info.

**Comment by fatcerberus** [comment]
> I do not have any flickering window buttons or similar.

Yeah, that seems to be a Windows thing.  I've seen Firefox do it a few times too, the X button starts flickering and no longer responds to clicks.  Kind of weird.

**Comment by SiegeLord** [comment]
Is this Win10?

**Comment by fatcerberus** [comment]
Win10 yes, but I remember this happening on 8.1 also.

**Comment by SiegeLord** [comment]
Tried a bit to reproduce this with Allegro's examples in a Win10 VM without success... I also tried it with minisphere, but being a VM, it can't use shaders, so that didn't go anywhere.

**Comment by fatcerberus** [comment]
That’s odd, I’ve always been able to run miniSphere in virtualbox, even with 3D acceleration turned off.

**Comment by fatcerberus** [comment]
After Christmas I’ll try to reproduce this on my end and post a stack trace.  Hopefully that will shed some light on what’s wrong.

**Comment by fatcerberus** [comment]
I've tracked this down to be a deadlock caused by an interaction between Allegro and a `printf` call.  The printf() hangs so the process never makes forward progress on the main thread.  Background threads (including the message pump) appear to continue functioning as normal.

Now that I have a lead I'll see if I can come up with a repro.

**Comment by beoran** [comment]
IIRC we are not supposed to even use printf to stdout on Windows in
windowed mode since the output buffer is limited. Maybe printing to a log
file will solve this.

**Comment by fatcerberus** [comment]
The standard “redist” build of miniSphere has all the printf’s disabled, this is for my “developer” build which is compiled as Console Application and therefore prints to the terminal like any other command line program.

In any case stdout is line-buffered by default when outputting to a terminal, so buffer size shouldn’t be an issue.  And even if it were, this happens at the *very first line printed* after display creation.

In my tests, issue only shows up if the display is out-of-focus (i.e. not the foreground window) when the first printf is encountered.  Once everything gets going it’s safe to background the window regardless of how much it prints afterwards, leading me to believe Allegro does something during initialization that blocks the printf and causes a deadlock.

It’s very possible that the hung printf is merely a symptom and the real root cause is something more general.  I won’t know until I run further experiments.

**Comment by beoran** [comment]
I see, then it would be interesting if you could reproduce the hung main
thread with something else but printf. As you say printf might be a red
herring here.
